cas-server-core-web/src/main/java/org/jasig/cas/web/flow/AbstractNonInteractiveCredentialsAction.java

```diff
@@ -68,17 +78,21 @@ protected final Event doExecute(final RequestContext context) {
         final String ticketGrantingTicketId = WebUtils.getTicketGrantingTicketId(context);
         final Service service = WebUtils.getService(context);
 
-        if (isRenewPresent(context)
-            && ticketGrantingTicketId != null
-            && service != null) {
+        if (isRenewPresent(context) && ticketGrantingTicketId != null && service != null) {
 
             try {
+                final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
```

Comment:
If I were to explain this in a more palatable way:

```
final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
                        this.authenticationObjectsRepository.getPrincipalElectionStrategy());
final AuthenticationTransaction transaction =
        this.authenticationObjectsRepository.getAuthenticationTransactionFactory().get(credential);
this.authenticationObjectsRepository.getAuthenticationTransactionManager().handle(transaction,  builder);
final AuthenticationContext authenticationContext = builder.build(service);
```

- Create a local authentication context at the start of the flow. Again, remember that there is only a SINGLE flow. One way to submit credentials. In your authentication context, only ONE thing happens. No more. 

- Obtain the credentials, and initiate an authentication transaction. The transaction will pre-vet the credentials. 

- Ask the Transaction Manager to process that transaction. 

(You may optionally repeat same 2-3, once we add subflows, and other ways to obtain credentials.  Again, remember that presently there is only a SINGLE flow. One way to submit credentials. In your authentication context, only ONE thing happens)

4. When you are satisfied, you finalize/build the context and instruct CASImpl to deal with it. 

Now, I think the question here is: do we have a proliferation of objects/components? Are we trying to create way too many abstractions? Possibly. Here is where the idea for these abstractions comes from:

- An `AuthenticationContext` is the playing ground. It's local to every request, from start to finish. It's as if you go to the gym, and you reserve the gym for a couple of hours from when you enter until when you exit.  If that makes sense, then we need something to build this context for us. Thus `AuthenticationContextBuilder`. 

- Once you are in the gym, you pass the front desk and present your credentials by swiping a card into a reader. The clerk will take your credentials, process them and gives you a pass to the gym. You enter the general hallway. 

The reader is the `AuthenticationTransactionManager`.
The act of you submitting credentials and swipping it is an `AuthenticationTransaction`
The clerk is the `AuthenticationManager`.
Your pass to the gym is an `Authentication` object. 

- Once you aren in the general hallway, you attempt to enter the basketball court. Another clerk stands at the door. You present your team membership card by swiping a card into a reader. The clerk will take your credentials, process them and gives you a pass to the gym. You enter the court. 


The reader is the `AuthenticationTransactionManager`.
The act of you submitting credentials and swipping it is an `AuthenticationTransaction`
The clerk is the `AuthenticationManager`.
Your pass to the gym is an `Authentication` object. 

Note that your credentials are different this time. 

- You are then ready to exit the gym. The gym (AuthenticationContext) knows this about you:

1. You presented credentials to the first desk. Were successful. We know who you are, your address, your name, your email, etc.
2. You presented credentials to the clerk at the court door. Were successful. We know you play basketball.

So the gym will provide a comprehensive view of ALL of your activity and ALL of who you are. That's where the finalized AuthenticationContext comes in. 
And you do this in the present patch via `AuthenticationContextBuilder#build()`

Note that again, in the present patch you do not enter the basketball court. There is absolutely no way for you to that via CAS. The act of submitting credentials takes place ONCE, and it's the only thing you can do. No SWF involvement; no need to record what you did, for the time being.